x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizeBlockHash.java

```diff
@@ -200,20 +200,45 @@ private Block buildIntermediateBlock() {
         }
         int positionCount = categorizer.getCategoryCount() + (seenNull ? 1 : 0);
         // We're returning a block with N positions just because the Page must have all blocks with the same position count!
-        return blockFactory.newConstantBytesRefBlockWith(serializeCategorizer(), positionCount);
+        var serializedCategorizer = serializeCategorizer();
+        return blockFactory.newConstantBytesRefBlockWith(
+            serializedCategorizer.bytesRef(),
+            positionCount,
+            serializedCategorizer.preAdjustedBytes()
+        );
     }
 
-    BytesRef serializeCategorizer() {
-        // TODO: This BytesStreamOutput is not accounted for by the circuit breaker. Fix that!
-        try (BytesStreamOutput out = new BytesStreamOutput()) {
+    record SerializedCategorizer(BytesRef bytesRef, long preAdjustedBytes) {}
+
+    /**
+     * Serializes the categorizer state into a BytesRef.
+     * <p>
+     *     The BytesRef is accounted for in the breaker, and its accounted value is returned too for future adjustments.
+     * </p>
+     * <p>
+     *     If the BytesRef is goind to be added to somewhere that will adjust the breaker, the pre-adjusted should be then
+     *     subtracted from the breaker.
+     * </p>
+     */
+    SerializedCategorizer serializeCategorizer() {
+        Long preAdjustedBytes = null;
+        try (ReleasableBytesStreamOutput out = new ReleasableBytesStreamOutput(blockFactory.bigArrays())) {
```

Comment:
Would it be crazy to make a `StreamOutput` that can serialize directly into the builder? I feel like if you added it as a private inner class to `BytesRefArray` you could vend it right from there and it could use the `bytes` member as the target for the bytes. And when it's closed you add the bytes to the builder ala `append` - except instead of copying the bytes you just record the end you reached.

This feels like it could almost be less code than this. Almost. More tests I guess.

But you could share this with some other places. And it removes the need for a copy entirely. So less accounting.